Added GLFW build support with -D BUILD_GLFW

teaganc opened the pull request:
Pretty simple, if you define BUILD_GLFW when compiling it should build with GLFW now instead for tut 6,7,8 and beyond. 

The only real change I made was changing the PLATFORM_SURFACE_EXTENSION_NAME solution to a function call defined in each platform specific file.

Other than that added a Window_glfw.cpp file and made the necessary changes in Window.h and Platform.h.

One other thing to note is I changed the MSVC files by hand because I didn't have it installed so you may just want to double check I didn't mess it up. Also I changed #include <Windows.h> to #include <windows.h> because when I am cross compiling on Linux it is necessary. 

If you want to make any changes be my guest, just thought GLFW support would be nice now that it shouldn't be any work to maintain.

Thanks.

Niko40 commented:
Awesome. Thank you. :)
All the MSCV files work perfectly.

It's all good but I'm going to request a few changes to your code before including it in my repository, could you do this? Also I just finished reviewing my own code and I updated it in a few places, including upping the used SDK to version 1.0.13.0 for all projects, which depricated a few lines of code around the solution ( device layers ) so I'd rather merge manually myself after you're done with the code. :)

Things I want changed before including it in my repository:
- Platform.h file: extern void AddRequiredPlatformInstanceExtensions(...), remove extern keyword, let me know if you think it should stay.
- Remove the AddRequiredPlatformInstanceExtensions() definitions from the Window_<platform>.h files and create a Platform.cpp file where you should include those definitions.
- glfwInit should have it's own place in the Renderer constructor, glfwTerminate isn't called it seems so add a call to that in the Renderer destructor. GLFW is a little bit special in this end and you can just add a #if defined (<glfw>) in the Renderer constructor and destructor directly. ( There's no check for it but there should only exist one Renderer at any one point in time so those functions should be called there ).
- All "#ifdef"s should be renamed to "#if defined()" for consistency.
- Change VK_USE_PLATFORM_GLFW_KHR into USE_FRAMEWORK_GLFW. This is because Vulkan can't use this information so we shouldn't use VK_ prefix, it's not an extension to vulkan so we shouldn't use _KHR suffix either, it's not even a platform but rather a framework.

Could you elaborate: "Also I changed #include to #include because when I am cross compiling on Linux it is necessary."
I might change some other things as well after you have done these changes.

Would this be alright to you?

teaganc commented:
That all makes a lot of sense to me.


I'll just update Tutorial 8, I suppose, if you want to manually add the changes (just let me know if you'd like me to do 6 and 7 also though.)

If you wanted to add some sort of function call along the lines of InitFramework and DeInitFramework in Platform.h instead of the #if in the Renderer Constructor and Deconstructor that might be more future proof. I could do that but for now I will just implement as you recommended.


For the #include <windows.h> thing, the header on a lot of systems is windows.h not Windows.h. For MSVC the capitalization doesn't matter but some compilers get mad and won't compile ( MinGW being an example ).



Note: The only effected files should now be Platform.h, Platform.cpp, Renderer.cpp, Window.h and Window_glfw.cpp


Thanks

________________________________
From: Niko40 <notifications@github.com>
Sent: May 26, 2016 5:33 PM
To: Niko40/Vulkan-API-Tutorials
Cc: teaganc; Author
Subject: Re: [Niko40/Vulkan-API-Tutorials] Added GLFW build support with -D BUILD_GLFW (#15)


Awesome. Thank you. :)
All the MSCV files work perfectly.

It's all good but I'm going to request a few changes to your code before including it in my repository, could you do this? Also I just finished reviewing my own code and I updated it in a few places, including upping the used SDK to version 1.0.13.0 for all projects, which depricated a few lines of code around the solution ( device layers ) so I'd rather merge manually myself after you're done with the code. :)

Things I want changed before including it in my repository:

  *   Platform.h file: extern void AddRequiredPlatformInstanceExtensions(...), remove extern keyword, let me know if you think it should stay.
  *   Remove the AddRequiredPlatformInstanceExtensions() definitions from the Window_.h files and create a Platform.cpp file where you should include those definitions.
  *   glfwInit should have it's own place in the Renderer constructor, glfwTerminate isn't called it seems so add a call to that in the Renderer destructor. GLFW is a little bit special in this end and you can just add a #if defined () in the Renderer constructor and destructor directly. ( There's no check for it but there should only exist one Renderer at any one point in time so those functions should be called there ).
  *   All "#ifdef"s should be renamed to "#if defined()" for consistency.
  *   Change VK_USE_PLATFORM_GLFW_KHR into USE_FRAMEWORK_GLFW. This is because Vulkan can't use this information so we shouldn't use VK_ prefix, it's not an extension to vulkan so we shouldn't use _KHR suffix either, it's not even a platform but rather a framework.

Could you elaborate: "Also I changed #include to #include because when I am cross compiling on Linux it is necessary."
I might change some other things as well after you have done these changes.

Would this be alright to you?

-
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/Niko40/Vulkan-API-Tutorials/pull/15#issuecomment-222023954>, or mute the thread<https://github.com/notifications/unsubscribe/ASVxp-VRKQ1dLskQm9z_tjAZUVeY8u4gks5qFi3AgaJpZM4InxXJ>.

Niko40 commented:
I can continue from here, thank you. :)
I will change all uppercase Window.h includes into lower case ones and I think I'll create a function I can initialize the GLFW with, it would be shared with platforms just in case so I don't know what to call it yet, might just call it InitPlatform because it's shared with platforms.

I'll add a list of contributors to each file, do you want your name or user name ( teaganc ) in there? If you want to add a link to your webpage I can also include that.

teaganc commented:
Awesome, InitPlatform sounds about right. My name is Teagan Chouinard; you can put in what ever you see fit.


Other than that, keep up the good work. It is appreciated :)

________________________________
From: Niko40 <notifications@github.com>
Sent: May 27, 2016 10:13 AM
To: Niko40/Vulkan-API-Tutorials
Cc: teaganc; Author
Subject: Re: [Niko40/Vulkan-API-Tutorials] Added GLFW build support with -D BUILD_GLFW (#15)


I can continue from here, thank you. :)
I will change all uppercase Window.h includes into lower case ones and I think I'll create a function I can initialize the GLFW with, it would be shared with platforms just in case so I don't know what to call it yet, might just call it InitPlatform because it's shared with platforms.

I'll add a list of contributors to each file, do you want your name or user name ( teaganc ) in there? If you want to add a link to your webpage I can also include that.

-
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/Niko40/Vulkan-API-Tutorials/pull/15#issuecomment-222188731>, or mute the thread<https://github.com/notifications/unsubscribe/ASVxp8cLBpb1dAWF24dKESYf7jzW8Q1Pks5qFxhEgaJpZM4InxXJ>.

Niko40 commented:
Okay. :)
One last thing I should ask. Are you okay with the CC0 licence? Even though I will add contributor lists these sources will be ultimately public domain.

If you are I'll upload the files shortly. :)

teaganc commented:
Yeah the code is yours to do with as you please. CC0 is always good for tutorials.

On May 28, 2016 2:26 PM, Niko40 <notifications@github.com> wrote:

Okay. :)
One last thing I should ask. Are you okay with the CC0 licence? Even though I will add contributor lists these sources will be ultimately public domain.

If you are I'll upload the files shortly. :)

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/Niko40/Vulkan-API-Tutorials/pull/15#issuecomment-222328157>, or mute the thread<https://github.com/notifications/unsubscribe/ASVxp53cRYJ6ea5AL8EHdbdyaul8wAh3ks5qGKUHgaJpZM4InxXJ>.

Niko40 commented:
Okay. The code is in master, thanks again for your help. :)
